File: Top200/114-Flatten.java
package Top200;

import java.util.Stack;

class Flatten {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode (int x) {
            val = x;
        }
    }

    // 类似莫里斯的方式
    public void flatten(TreeNode root) {
        TreeNode curr = root;

        while (curr != null) {
            if (curr.left != null) {
                TreeNode pre = curr.left;
                
                // 找到 curr.right 的前驱节点
                while (pre.right != null) {
                    pre = pre.right;
                }

                pre.right = curr.right;
                curr.right = curr.left;
                curr.left = null;

                curr = curr.right;
            } else {
                curr = curr.right;
            }
        }
    }

    public TreeNode pre = null;

    public void flatten2(TreeNode root) {
        if (root == null) {
            return;
        }

        flatten2(root.right);
        flatten2(root.left);

        root.right = pre;
        root.left = null;

        pre = root;
    }

    public void flatten3(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode curr = root;
        TreeNode pre = null;

        while (curr != null || !stack.isEmpty()) {
            if (curr != null) {
                if (curr.right != null) {
                    stack.push(curr.right);
                }

                pre = curr;
                curr.right = curr.left;
                curr.left = null;
                curr = curr.right;
                
            } else {
                TreeNode temp = stack.pop();
                pre.right = temp;
                pre =  temp;
                curr = temp;
            }
        }
    }
}

File: Top200/114-Flatten.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	if root == nil {
		return
	}
	var prev *TreeNode = nil

	flattenRecur(root, &prev)
}

func flattenRecur(root *TreeNode, prev **TreeNode) {
	if root == nil {
		return
	}

	flattenRecur(root.Right, prev)
	flattenRecur(root.Left, prev)

	root.Right = *prev
	root.Left = nil

	*prev = root
}

func flatten2(root *TreeNode) {
	if root == nil {
		return
	}

	flattenRecur2(root)
}

func flattenRecur2(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	right := flattenRecur2(root.Right)
	left := flattenRecur2(root.Left)

	if left == nil {
		root.Right = right
		return root
	}

	curr := left

	for curr.Right != nil {
		curr = curr.Right
	}

	curr.Right = right

	root.Right = left
	root.Left = nil

	return root
}

func flattenRecur3(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	left := flattenRecur3(root.Left)
	right := flattenRecur3(root.Right)

	if left == nil {
		root.Right = right
		return root
	}

	curr := left

	for curr.Right != nil {
		curr = curr.Right
	}

	curr.Right = right

	root.Right = left
	root.Left = nil

	return root
}
